### ISSUE TITLE
[Proposal] Asyncio integration + Supervisor refactor

### ISSUE THREAD

**Issue by JackUrb** [facebookresearch/Mephisto#514]
# Motivation
As we move into higher levels of scaling, we're running into a tradeoff between needing to achieve higher throughput with multithreading/multiprocessing and running into difficult-to-debug race conditions. By adding more locking, we restrict code paths, and this doesn't long-term help us achieve multiprocessing where possible.

As a midpoint, this proposal is such that we break up the Supervisor class into clearer encapsulated components, and upgrade to using `asyncio` for clearer code execution pathways in the process.

# Rough Plan
###Long-term Multiprocessing Ideals:
- Run main threads without being blocked by task worlds
- Not be blocked by the server and responses
- Retain Monitoring in multiprocessing down the road
- Retain single point of entry
- Still works with a single thread for current design

Plan to have 1 point of entry still, the main operator call. API doesn’t really change in this way. After initialization, it splits into threads of three core categories: IO, Mephisto, and Task. IO covers everything coming in from the server, Mephisto covers monitoring and delegation, and Task covers runtime for a specific Unit or Assignment instance. 

The supervisor currently acts as all of the message queue manager, the thread launcher, and the status centralization. This is somewhat opaque, and becomes worse if we try to simply port to multiprocessing. 

In the IO pool is the MessageHandler. It is responsible for monitoring all of the Channels, and enqueueing incoming and outgoing messages. It is also responsible for monitoring agent liveliness and status.

In the Task pool, we need to have workers that receive ready agent-unit combinations from a queue and then launch and execute.

Process priority is to first split Mephisto and IO out from Tasks, then give IO own process, then split remaining processes to more Tasks. Future work can let people directly configure process trees.

### Requirements:
- Now:
  - Split up supervisor into IO message queue, thread launching, and status management and monitoring. 
  - Update downstream functions to be asynchronous where possible.
  - Ensure cleanup collects all relevant processes.
- Later:
  - Create Multiprocessing Agent, TaskRunner, and SharedTaskState
  - Make Operator handle delegating and launching things in processes, if configured to launch using multiprocessing enabled.

### Risks:
- Multiprocessing may have reduced feature set:
  - Ease of passing functions over between main process and others is unclear
  - Initialization of unit state is best if not dependent on info from the original process
- Multiprocessing may introduce locks
  - Interference with DB writes and accesses
- Additional complexity:
  - Message passing from incoming messages through to task worlds dynamically is unclear

# Asyncio Implementation Steps
### Splitting the supervisor:
- [x] MessageHandler #577 
  - [x] Channel handling
  - [x] Incoming and outgoing message
  - [x] Requests and statuses
- [x] Status handling (new Supervisor): #634 
  - [x] Incoming status from frontend
  - [x] Updating agent statuses
- [x] Task Manager #634 
  - [x] Getting initialization of data
    -  This should now be prepared by the TaskRunner, rather than elsewhere
  - [x] Execution of tasks
  - [x] Unit selection and registration
### Downstream async changes:
- [x] TaskRunner needs to be entirely revamped
- [ ] Hooks should be async
- [ ] Get_valid_units_for_worker should be updated as well

**Comment by pringshia** [comment]
Do we have someone else with async Python experience who may also be able to shed thoughts on this proposal? Perhaps we can bribe someone from the ParlAI team, e.g. @stephenroller :)

**Comment by stephenroller** [comment]
I'm going to suggest @moyapchen instead. She may not be aware obscure behaviors of python's asyncio, but she's highly appropriate for a design level discussion

**Comment by moyapchen** [comment]
Yeah, I don't have background for python's asyncio, but if it functions similarly to async/await in C++11, I can give at least a little commentary.

Initial thoughts from skimming the above:
1. At a high level, the details written there sound about right, but a quick picture of the expected relationship between different components such that it's explicit would be great. 
2. For shared resources, a previous team used to get around the locking issue using an EventQueue around the resource. It wouldn't necessarily prevent using locks ever (since the EventQueue has internal locks for putting in and processing events) but did seem to simplify reasoning about locks some. (Not sure if it fully applies here since that was a system where dropping events were fine)

**Comment by JackUrb** [comment]
Closed in #650, releasing in 1.0
